### DISCUSSION TITLE
Query, Mutation and Subscriptions [ALL] over websockets

### DISCUSSION THREAD

**Discussion by paulm17** [urql-graphql/urql#2190] [Q&A] [answered]
> Similar to this question: https://spectrum.chat/apollo/general/query-mutation-subscriptions-over-one-websocket-connection~a324fef0-d1b1-45f7-a784-449ddf71ab08

I'm currently investigating graphql for a possible migration of an existing development using centrifugo, which is a websocket server writting in golang.

Will we ever see the ability to do queries and mutations using websockets, instead of http?

I can understand initial requests before websockets have bootstrapped, but once the connection is in play.  Websockets have a much lower overhead.

Thanks

**Comment by JoviDeCroock** [comment, marked as answer]
It is already possible, here is an [example repo](https://github.com/andyrichardson/graphql-http-vs-ws/blob/master/frontend/src/benchmark.ts#L25) leveraging the `enableAllOperations` property of our `subscriptionExchange`

**Reply by paulm17** [reply]
Thanks for the reply.  Unfortunately I haven't been able to get this to work.

I'm using NextJS, but opted not to use SSR.

I've got the following client:

```
const wsClient = createWSClient({
    url: "wss://dgraph.platformdev.com/graphql",
  })

  const client = createClient({
    url: "https://dgraph.platformdev.com/graphql",
    exchanges: [
      ...defaultExchanges,
      subscriptionExchange({
        enableAllOperations: true,
        forwardSubscription: (operation) => ({
          subscribe: (sink) => ({
            unsubscribe: wsClient.subscribe(operation, sink),
          }),
        }),
      }),
    ],
    fetchOptions: () => {
      return { headers: {} }
    },
  })
```

And using this via the provider:

```
<Provider value={getGraphQLClient()}>
    <Component {...pageProps} />
</Provider>
```

Looking at the network tab, all my hook requests are still using http.

I've tried to emulate what you have linked to, using client.

```
const client = useClient()

const data = client
    .query(UserDocument, { xid: userId })
    .toPromise()
    .then((result) => {
      console.log(result)
    })
```

Still using http for the requests.

Can you clarify whether hooks can use websockets?

Also can you please give a clear example which should rely on wss.  As I am unable to match what's in the link you provided.

Thanks

**Reply by JoviDeCroock** [reply]
In your defaultExchanges there is a fetchExchange which will use http, you might want to only add `dedup` and `cache` as exchanges

```js
import { dedupExchange, cacheExchange, subscriptionExchange, createClient } from 'urql';

const wsClient = createWSClient({
    url: "wss://dgraph.platformdev.com/graphql",
  })

  const client = createClient({
    url: "https://dgraph.platformdev.com/graphql",
    exchanges: [
      dedupExchange,
      cacheExchange,
      subscriptionExchange({
        enableAllOperations: true,
        forwardSubscription: (operation) => ({
          subscribe: (sink) => ({
            unsubscribe: wsClient.subscribe(operation, sink),
          }),
        }),
      }),
    ],
    fetchOptions: () => {
      return { headers: {} }
    },
  })
```

**Reply by paulm17** [reply]
Thank you for the quick follow up!

I now have websockets working.  It is complaining about something, so it's not connecting.  Under apollo it seemed to be fine.

But I'll troubleshoot this my side.

Many thanks!

**Reply by paulm17** [reply]
I changed the websocket implementation to subscriptions-transport-ws and I don't get any issues.

**Reply by JoviDeCroock** [reply]
hmm yeah, I've heard this before where `graphql-ws` is incompatible with the backend version of `subscriptions-transport-ws` 😅

**Reply by tornqvist** [reply]
I'm having trouble with his approach, getting the error `Expected operation of type "subscription" but found "query"`.

It appears that a check in `createRequestOperation` introduced in [#1829](https://github.com/urql-graphql/urql/commit/1773e6ee509e514da9b34cb023d4d1b7a43cbe93#diff-858339789c798051d9c13620bffce8ce860e7ad8ba18dba1c8711eb1f7d51a55R299) will throw if trying to send a `query` or `mutation` over WebSockets before ever getting to the `enableAllOperations`-check. Is this intended behavior?

Ironically it does work in production since the check is only done in development.

**Reply by kitten** [reply]
@tornqvist: That error is specifically related to mismatching operations and APIs, e.g. passing a query to `useSubscription` / `client.executeSubscription`.

The operation kind and API used must always match.

**Reply by tornqvist** [reply]
Oh, that makes sense. So that means I should still use `client.executeQuery` (or `createQuery`, as I'm using svelte) and just not have a fetchExchange in there to have the subscriptionExchange handle the query? 

Unfortunately, I need to have fetchExchange since it's only certain queries I want to execute over WebSocket. I've tried to implement an intermediate exchange to determine whether to use fetchExchange or forward to subscriptionExchange but am getting a `forward() must only be called once in each Exchange.` error, just like https://github.com/urql-graphql/urql/discussions/3559

How would one go about doing this?

```js
function maybeFetchExchange(params) {
  const fetch = fetchExchange(params)
  return (operations$) => {
    const toWebSocket = pipe(
      operations$,
      filter((operation) => operation.context.useWebSocket),
      params.forward
    )

    const toFetch = pipe(
      operations$,
      filter((operation) => !operation.context.useWebSocket),
      fetch
    )

    return merge([toFetch, toWebSocket])
  }
}
```
